src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs

```diff
@@ -136,6 +160,37 @@ private void OnRpcDisconnected(object sender, JsonRpcDisconnectedEventArgs e)
             Disconnected();
         }
 
+        private static async Task<TResult> RetryRemoteCallAsync<TInput, TException, TResult>(
+            Func<TInput, CancellationToken, Task<TResult>> funcAsync,
+            TInput input,
+            TimeSpan timeout,
+            CancellationToken cancellationToken) where TException : Exception
+        {
+            const int retry_delayInMS = 50;
+
+            var start = DateTime.UtcNow;
+            while (DateTime.UtcNow - start < timeout)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                try
+                {
+                    return await funcAsync(input, cancellationToken).ConfigureAwait(false);
+                }
+                catch (TException)
```

Comment:
parameterizing TExceptoin seems pretty strange.  Can you explain the different types of exception behavior you're trying to deal with?

Reply:
> since I don't want it to catch all exceptions blindly and retry. I want only specific exception to be captured and all others don't touch.

Sure.  I just feel that the flow control is wonky, and we need comments.  For example, it's possible for the exception to be caught and have the cancellaiton token *not* be triggered, thus having the code 'fall through' out of the catch.  This is definitely by design.  But it also merits some documentation because it might not be clear to others that that's hte behaivor we want.

For example, someone may think "this should be translating all of these exception types to a cancellation exception" not realizing that sometimes you fall through and sometimes you don't.

Reply:
Or, in other words, this construct is extremely rare in roslyn.  extremely rare constructs warrant more clarity through documentation :)